R/write_output.R

```diff
@@ -0,0 +1,336 @@
+        "tasks",
+        task_id,
+        "model.rds"
+      )
+      saveRDS(fit, model_path)
+      cli::cli_alert_success("Wrote model to {.path {model_path}}")
+
+      # Write model run metadata
+      metadata_path <- file.path(
+        output_dir,
+        job_id,
+        "tasks",
+        task_id,
+        "metadata.json"
+      )
+      jsonlite::write_json(metadata, metadata_path, pretty = TRUE)
+      cli::cli_alert_success("Wrote metadata to {.path {metadata_path}}")
+    },
+    error = function(cnd) {
+      # Downgrade erroring out to a warning so we can catch and log
+      cli::cli_warn(
+        "Failure writing outputs",
+        parent = cnd,
+        class = "no_outputs"
+      )
+    }
+  )
+
+  invisible(NULL)
+}
+
+#' Create output directory structure for a given job and task.
+#'
+#' This function generates the necessary directory structure for storing output
+#' files related to a job and its tasks, including directories for raw samples
+#' and summarized quantiles.
+#'
+#' @param output_dir String. The base output directory path.
+#' @param job_id String. The identifier for the job.
+#' @param task_id String. The identifier for the task.
+#'
+#' @return The path to the base output directory (invisible).
+#' @export
+write_output_dir_structure <- function(output_dir, job_id, task_id) {
+  # Top level
+  dir.create(output_dir,
+    showWarnings = FALSE
+  )
+  # Second level
+  dir.create(file.path(output_dir, job_id),
+    showWarnings = FALSE
+  )
+  # Third level
+  dir.create(file.path(output_dir, job_id, "tasks"),
+    showWarnings = FALSE
+  )
+  dir.create(file.path(output_dir, job_id, "samples"),
+    showWarnings = FALSE
+  )
+  dir.create(file.path(output_dir, job_id, "summaries"),
+    showWarnings = FALSE
+  )
+
+  # Fourth level
+  dir.create(file.path(output_dir, job_id, "tasks", task_id),
+    showWarnings = FALSE
+  )
+
+  invisible(output_dir)
+}
+
+#' Extract posterior draws from a Stan fit object.
+#'
+#' This function extracts posterior draws for specific parameters from a Stan
+#' fit object and prepares a fact table containing unique date-time-parameter
+#' combinations for further merging.
+#'
+#' @param fit A Stan fit object with posterior estimates.
+#'
+#' @return A list containing two elements: `stan_draws` (the extracted draws in
+#' long format) and `fact_table` (a table of unique date-time-parameter
+#' combinations).
+#' @noRd
+extract_draws_from_fit <- function(fit) {
+  # Step 1: Extract unique date-time-parameter combinations
+  fact_table <- fit[["estimates"]][["samples"]][,
+    c("date", "time", "parameter"),
+    with = FALSE
+  ]
+  fact_table <- stats::na.omit(unique(fact_table))
+
+  # Step 1.1: Add corresponding 'obs_cases' rows for 'latent_cases' dates
+  obs_fact_table <- fact_table[
+    fact_table[["parameter"]] == "imputed_reports",
+  ]
+  # The EpiNow2 summary table has the variable `imputed_reports`
+  # for nowcast-corrected cases, but not `obs_reports` for right-
```

Comment:
If you wanted it `reports` would be expected cases without right truncation or error. I can imagine you do want it though

Reply:
https://github.com/epiforecasts/EpiNow2/blob/2e1aa86598f13364c1af05a5a83b3ac7790db41a/inst/stan/estimate_infections.stan#L154

Reply:
Ok summarizing my conversation with @seabbs here:

EpiNow2 estimates the incident cases $\hat y_{td}$ for timepoint $t \in \{1, ..., T\}$ and delay $d \in \{1, ..., D\}$ where $D \le T$. The delay $d$ moves inversely to timepoints, so $d = T - t + 1$.

The observed data vector of length $T$ is $y_{td} \in W$. We supply a nowcasting correction PMF $\nu$ for the last $D$ timepoints where $\nu_d \in [0, 1],$ and $\sum_{\nu=1}^Dp_d = 1$. We also have some priors $\Theta$.

We use EpiNow2's generative model $f(y, \nu, \Theta)$.

EpiNow2 is a forward model that produces an expected nowcasted case count for each $t$ and $d$ pair: $\gamma_{td}$. 
 It applies the nowcasting correction $\nu$ to the last $D$ timepoints of $\gamma$ to produce the expected right-truncated case count $\hat y$. Note that these _expected_ case counts (with and without right-truncation) don't have observation noise included.

We can apply negative binomial observation noise using EpiNow2's estimate of the negative binomial overdispersion parameter $\hat \phi$ and the expected case counts. The posterior predictive distributions of nowcasted case counts is $\tilde \gamma \sim \text{NB}(\hat \gamma, \hat \phi)$. The posterior predicted right-truncated case count is $\tilde y \sim \text{NB}(\hat y, \hat \phi)$.

We can get 3 of these 4 quantities pre-generated from the returned EpiNow2 Stan model:

- $\hat \gamma$: The expected nowcasted case count is `reports[t]`
- $\hat y$: The expected right-truncated case count is `obs_reports[t]`
- $\tilde \gamma$: The posterior-predicted nowcasted case count is `imputed_reports`
- $\tilde y$: The posterior-predicted right-truncated case count isn't returned by EpiNow2. We could (and probably should) generate it ourselves in post-processing by passing each draw $\hat \gamma_{td}$ to `rnbinom()` with $\hat \phi$.

Sam does that sound correct to you? @kgostic and @athowes thoughts on naming based on this? I'm happy to make a new issue for generating $\tilde y$ in post-processing if we think we should (for scoring maybe?).

Reply:
yes sounds correct to me. Two points. 

1. Rather than generating yourself you could ask for 4. as an `EpiNow2` output. It seems to me there is a case for that as a general thing but its probably a do later vs for this feature
2. The model is actually for $T +h$ where $h$ is the forecast horizon but this doesn't impact anything you have written except that the actual dimensions are: `reports[t+h]`, `obs_reports[t]`, and `imputed_reports[t+h]`.

Reply:
Ok I need to think this through a bit more but I think that this method of timepoint -> date extraction isn't quite lined up:

https://github.com/CDCgov/cfa-epinow2-pipeline/blob/e0729f97987b78b12997fd337cbe2edb21b9c66f/R/write_output.R#L145-L161

Reply:
Ok post-lunch I now think this is ok because the impact on `imputed_reports` is to increase the maximum dimension $T$ but not the alignment of the vector to the dates. So it should line up ok, but I may need to expand the maximum of the vector to avoid NAs.

Reply:
https://github.com/CDCgov/cfa-epinow2-pipeline/pull/41/commits/f92ccb859fbaf456a84e8cf60572d5bee2346291